Allow emojis with variant selectors to be validated

abhiklodh opened the pull request:
<!--
Before contributing (PLEASE READ!)

⚠️ If your contribution is more than a few lines of code, then prior to starting to code on it please post in the issue saying you want to volunteer, then wait for a positive response. And if there is no issue for it yet, create it first.

This helps make sure:

  1. Two people aren't working on the same thing
  2. This is something Streamlit's maintainers believe should be implemented/fixed
  3. Any API, UI, or deeper architectural changes that need to be implemented have been fully thought through by Streamlit's maintainers
  4. Your time is well spent!

More information in our wiki: https://github.com/streamlit/streamlit/wiki/Contributing
-->

## 📚 Context

Emojis show up as invalid if they are prefixed or postfixed with a variant selector character. This is a hidden character that is used as variant selector for emojis (more information [here](https://stackoverflow.com/questions/38100329/what-does-u-ufe0f-in-an-emoji-mean-is-it-the-same-if-i-delete-it)).

- What kind of change does this PR introduce?

  - [X] Bugfix
  - [ ] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- Wrote a function `clean_emoji` to remove hidden unicode and return a sanitised emoji if possible.
- Added logic to `validate_emoji` to handle emojis that have hidden unicode with the help of `clean_emoji`.

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

**Revised:**

```
st.success("This is a success message", icon="🚨") # Works fine
st.success("This is a success message", icon="️🚨") # Works fine
st.success("This is a success message", icon="\U0000FE0F\U0001F44D\U0001F3FD\U0000FE0F") # Works fine
st.success("This is a success message", icon="\U0000FE0F") # Throws an error
st.success("This is a success message", icon="\U0000FE0F\U0001F6A8\U0000FE0F") # Works fine
st.warning("This is a warning", icon="🚨") # Works fine
st.warning("This is a warning", icon="️🚨") # Works fine
st.error("This is an error", icon="🚨") # Works fine
st.error("This is an error", icon="️🚨") # Works fine
st.info("This is a purely informational message", icon="🚨") # Works fine
st.info("This is a purely informational message", icon="️🚨") # Works fine
```

**Current:**

```
st.success("This is a success message", icon="🚨") # Works fine
st.success("This is a success message", icon="️🚨") # Throws an error
st.success("This is a success message", icon="\U0000FE0F\U0001F44D\U0001F3FD\U0000FE0F") # Throws an error
st.success("This is a success message", icon="\U0000FE0F") # Throws an error
st.success("This is a success message", icon="\U0000FE0F\U0001F6A8\U0000FE0F") # Throws an error
st.warning("This is a warning", icon="🚨") # Works fine
st.warning("This is a warning", icon="️🚨") # Throws an error
st.error("This is an error", icon="🚨") # Works fine
st.error("This is an error", icon="️🚨") # Throws an error
st.info("This is a purely informational message", icon="🚨") # Works fine
st.info("This is a purely informational message", icon="️🚨") # Throws an error
```

## 🧪 Testing Done

- [ ] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: Closes #5564 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

abhiklodh committed:
Added logic to handle emoji with hidden unicode

abhiklodh committed:
Add method to sanitise emoji with hidden unicode

abhiklodh committed:
Merge branch 'streamlit:develop' into invalid-emoji

kajarenc commented:
Hello @abhiklodh, thanks for your contribution

I think we should not modify the emoji, and only "normalize" it for `is_emoji` check to keep the selector.

This comment could be helpful for you in case you want to work on this https://github.com/streamlit/streamlit/pull/5582#issuecomment-1287065330

abhiklodh committed:
Update string_util.py

abhiklodh committed:
Update alert.py

abhiklodh committed:
Simplified logic to prevent modifying emoji

abhiklodh committed:
Merge branch 'streamlit:develop' into invalid-emoji

abhiklodh commented:
> Hello @abhiklodh, thanks for your contribution
> 
> I think we should not modify the emoji, and only "normalize" it for `is_emoji` check to keep the selector.
> 
> This comment could be helpful for you in case you want to work on this [#5582 (comment)](https://github.com/streamlit/streamlit/pull/5582#issuecomment-1287065330)

Hey @kajarenc!
Thanks for the cleaner suggestion. I pushed my commit implementing the same and it works as intended.

kajarenc commented:
> > Hello @abhiklodh, thanks for your contribution
> > I think we should not modify the emoji, and only "normalize" it for `is_emoji` check to keep the selector.
> > This comment could be helpful for you in case you want to work on this [#5582 (comment)](https://github.com/streamlit/streamlit/pull/5582#issuecomment-1287065330)
> 
> Hey @kajarenc! Thanks for the cleaner suggestion. I pushed my commit implementing the same and it works as intended.

@abhiklodh could you please fix formatter error, and commit again to rerun CI:

```python
 def is_emoji(text: str) -> bool:
     """Check if input string is a valid emoji."""
-    return (text in ALL_EMOJIS) or ("\U0000FE0F" in text and text.replace("\U0000FE0F", "") in ALL_EMOJIS)
+    return (text in ALL_EMOJIS) or (
+        "\U0000FE0F" in text and text.replace("\U0000FE0F", "") in ALL_EMOJIS
+    )
```

abhiklodh committed:
Merge branch 'streamlit:develop' into invalid-emoji

abhiklodh committed:
Fix linting issues

vdonato commented on lib/streamlit/string_util.py:
```diff
@@ -43,7 +43,9 @@ def clean_text(text: "SupportsStr") -> str:
 def is_emoji(text: str) -> bool:
     """Check if input string is a valid emoji."""
-    return text in ALL_EMOJIS
+    return (text in ALL_EMOJIS) or (
+        "\U0000FE0F" in text and text.replace("\U0000FE0F", "") in ALL_EMOJIS
+    )
```
nit: I think it'd be equivalent and very slightly cleaner to write

```python
return text.replace("\U0000FE0F", "") in ALL_EMOJIS
```

since the `.replace` should be a noop in the case that `text in ALL_EMOJIS`

vdonato reviewed (commented):
Hi @abhiklodh,

Sorry for the slow turnaround time on reviews for this PR!

This change LGTM, but I have one small change that I suggested along with a request that a couple unit tests get added to verify that these changes work as expected.

There are already some tests for this function that live here: https://github.com/streamlit/streamlit/blob/f1193d88e59c8458761fa17076c65082c7b7a752/lib/tests/streamlit/string_util_test.py#L27
so adding another 2-3 cases to the parameterized list should be sufficient (copying some of the examples listed in the PR description should already be enough).

Once that's done, this PR should be good to go

abhiklodh committed:
Merge branch 'streamlit:develop' into invalid-emoji

abhiklodh committed:
Cleanup code following suggestions

abhiklodh committed:
Added 3 test cases to validate emojis with variant selectors

abhiklodh commented:
Hey @vdonato 
I added three test cases for the unit tests. 
https://github.com/abhiklodh/streamlit/blob/736b1093b5bf7f20c54a977d104c4a7c5ca28475/lib/tests/streamlit/string_util_test.py#L27-L41

And cleaned up this line too. https://github.com/abhiklodh/streamlit/blob/736b1093b5bf7f20c54a977d104c4a7c5ca28475/lib/streamlit/string_util.py#L44-L46
Let me know if there's anything else left to be done.

vdonato reviewed (approved):
This LGTM

Thanks @abhiklodh!
